Add StBadRunChecker

liuxiaoyuyuyu opened the pull request:
StBadRunChecker is used to check bad runs based on different STAR sub-systems for BESII data. BadRunSubSys.h is used to store the bad runs based on subsystems. It now has run19 19.6 GeV injection runs and TPC bad runs.

liuxiaoyuyuyu committed:
Add files via upload

marrbnl commented on StRoot/StRefMultCorr/BadRunSubSys.h:
```diff
@@ -0,0 +1,38 @@
+
+const Int_t run19_col_19p6_auau_range[2]={20056032,20093036};
+const Int_t nBadRun_run19_col_19p6_auau =374;
+const Int_t badrun_run19_col_19p6_auau_all[nBadRun_run19_col_19p6_auau]={20076040,20076045,20076048,20082063,20082060,20082066,20082065,20073074,20073076,20073071,20073072,20079006,20081002,20081004,20081007,20065060,20060025,20060022,20079009,20066078,20057007,20066073,20073002,20073006,20069032,20069033,20069030,20069031,20078001,20066008,20078007,20090024,20066001,20062011,20062010,20062012,20 ...
```
Judging from the way this array is used in the cxx file, I think there should not be any duplicated entries in the array. If this is true, please add a comment such that future users do not prepare the wrong format. At the moment, the injection runs and TPC bad runs do not overlap by design, as the TPC bad runs are determined after removing the injection runs. However, the bad run lists for different subsystems could in principle overlap, and one would need a dedicated macro or script to prepare the array from the lists of bad runs.

liuxiaoyuyuyu replied:
I have a python script for preparing the correct format, I will add it to the next pull request.

marrbnl commented on StRoot/StRefMultCorr/BadRunSubSys.h:
```diff
@@ -0,0 +1,38 @@
+const Int_t run19_col_19p6_auau_range[2]={20056032,20093036};
+const Int_t nBadRun_run19_col_19p6_auau =374;
+const Int_t badrun_run19_col_19p6_auau_all[nBadRun_run19_col_19p6_auau]={20076040,20076045,20076048,20082063,20082060,20082066,20082065,20073074,20073076,20073071,20073072,20079006,20081002,20081004,20081007,20065060,20060025,20060022,20079009,20066078,20057007,20066073,20073002,20073006,20069032,20069033,20069030,20069031,20078001,20066008,20078007,20090024,20066001,20062011,20062010,20062012,20 ...
+const Int_t badrun_run19_col_19p6_auau_sub[nBadRun_run19_col_19p6_auau][12]={{1,0,0,0,0,0,0,0,0,0,0,0},{1,0,0,0,0,0,0,0,0,0,0,0},{1,0,0,0,0,0,0,0,0,0,0,0},{1,0,0,0,0,0,0,0,0,0,0,0},{0,1,0,0,0,0,0,0,0,0,0,0},{1,0,0,0,0,0,0,0,0,0,0,0},{0,1,0,0,0,0,0,0,0,0,0,0},{1,0,0,0,0,0,0,0,0,0,0,0},{0,1,0,0,0,0,0,0,0,0,0,0},{0,1,0,0,0,0,0,0,0,0,0,0},{0,1,0,0,0,0,0,0,0,0,0,0},{1,0,0,0,0,0,0,0,0,0,0,0},{0,1,0,0,0 ...
```
Similar to the comment for line 31, a dedicated macro/script is needed to prepare this list. We need to upload those macros/scripts as well in another PR

Maybe you can use "isbadrun" instead of "badrun" to make it clear that the numbers in this array indicate whether a particular run from the "badrun_*_auau_all" array is bad or not for a particular subsystem. It took me a while to figure it out. Would be good to add a comment about this. You can also consider using Bool_t instead of Int_t

liuxiaoyuyuyu replied:
Will do

marrbnl commented on StRoot/StRefMultCorr/StBadRunChecker.cxx:
```diff
@@ -0,0 +1,317 @@
+    }
+    //return ( iter != mBadRun.end() ) ;
+    else{
+        int index = std::distance(mBadRun_all.begin(), iter);//index starts at zero
```
For this function to work, one needs to make sure there are no duplicated entries in "badrun_*_all" array

liuxiaoyuyuyu replied:
Yes, the python script can make sure there is no duplicated entries in "badrun_*_all" array

marrbnl commented on StRoot/StRefMultCorr/StBadRunChecker.cxx:
```diff
@@ -0,0 +1,317 @@
+    }
+}
+
+Bool_t StBadRunChecker::isInjection(const Int_t RunId){
```
Instead of implementing new codes, could you just call "isBadRun(RunId, Injection)"? Same comments for all other subsystem-specific functions.

liuxiaoyuyuyu replied:
Yes, you can just call isBadRun(RunId,injection). The subsystem-specific functions are indeed optional. I was thinking maybe some users prefer calling methods with explicit names instead of passing arguments to a method themselves, so they don't need to worry about typos? It doesn't hurt to keep both but I am also okay with removing all the subsystem-specific functions if you think it will cause confusion.

marrbnl replied:
I think it is a good idea to have those subsystem-specific functions. My comment was that you can use the master isBadRun() function for implementation, instead of writing somewhat duplicated codes. Something like: 
Bool_t StBadRunChecker::isInjection(const Int_t RunId){
return isBadRun(RunId,injection);
}

liuxiaoyuyuyu replied:
I see! Good idea!

marrbnl commented on StRoot/StRefMultCorr/StBadRunChecker.cxx:
```diff
@@ -0,0 +1,317 @@
+    //return ( iter != mBadRun.end() ) ;
+    else{
+        int index = std::distance(mBadRun_all.begin(), iter);//index starts at zero
+        int isbad=1;
```
If I understand correctly, you use "isbad=0" as the run being bad and "isbad=1" being good, which is a bit counter-intuitive. This is also opposite to the convention used in "badrun_*_sub".

liuxiaoyuyuyu replied:
You are right. I want to implement a "or" logic in this method, so it is easier to set isbad=0 for bad runs. I can change the name to "isgood" to avoid confusion.

marrbnl commented on StRoot/StRefMultCorr/StBadRunChecker.h:
```diff
@@ -0,0 +1,50 @@
+  Bool_t isBadRunBEMCTrigger(const Int_t RunId) ;
+  Bool_t isBadRunMTD(const Int_t RunId) ;
+  Bool_t isBadRunAnalysis(const Int_t RunId) ; 
+  Bool_t isBadRun(const Int_t RunId,TString mSys) ; 
```
Would be good to add a comment that "mSys" could be for a single sub-system or a combination of different sub-systems, such as "injection" or "tpc_etof_vpd", etc.

marrbnl commented on StRoot/StRefMultCorr/StBadRunChecker.h:
```diff
@@ -0,0 +1,50 @@
+  std::vector<Int_t> mRunRange;
+  //const TString mSubSysName[12]={"Injection","TPC","bTOFStatus","bTOFPID","eTOF","EPD","VPD","BEMCStatus","BEMCPID","BEMCTrigger","MTD","Analysis"};
+  TString mSubSysName[12]={"Injection","TPC","bTOFStatus","bTOFPID","eTOF","EPD","VPD","BEMCStatus","BEMCPID","BEMCTrigger","MTD","Analysis"};
+  TString mRun;
```
Add comments to mRun, mEnergy, mColMode, mSpecies about the code expects them to be look like, such as "run15, run19", "200, 19.6", "col, fxt", "auau, pau", etc. In this way, users know how to set them up.

liuxiaoyuyuyu replied:
Will do

liuxiaoyuyuyu committed:
StBadRunChecker update

Added a python script for updating BadRunSubSys.h. Added more comments. implemented suggestions from Rongrong.

nigmatkulov commented:
Hi @liuxiaoyuyuyu , please create a new branch (from main) and commit modifications related to StBadRunChecker there.
When it will be ready please let me know. I'll close this commit.
